### PR TITLE
Add optional static enum lint

### PR DESCRIPTION
**Pull request by jridgewell**
This adds an optional enum linter that ensures that an enum is only used in a statically DCE-able way. Enums opt-into this behavior by using the `_Enum` suffix in their name.

Depends on #36755
Partial for #36754

### CODE REVIEW

File: build-system/eslint-rules/enums.js [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+      }
+    }
+  }
+
+  /**
+   * @param {!Node} node
+   */
+  function checkStaticEnumUse(node) {
+    const {parent} = node;
+    // import { Enum } from '.'
```
Comment (samouri) [inline build-system/eslint-rules/enums.js:91]:
```suggestion
```

File: build-system/eslint-rules/enums.js [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    if (parent.type === 'CallExpression') {
+      const {arguments: args, callee} = parent;
+      if (args[0] === node) {
+        // isEnumValue(Enum, value)
+        if (callee.type === 'Identifier') {
+          const {name} = callee;
+          if (
+            name === 'isEnumValue' // ||
```
Comment (samouri) [inline build-system/eslint-rules/enums.js:124]: either:
- a) remove comments? same for lower down ones
- b) add a TODO line w/ a small explanation of why this is here

**Review by samouri — COMMENTED** [review]
IMO this is not worth disabling for `build-system`. In fact, build-system should likely never use `@enum` since it isn't compiled and therefore the annotation has no meaning

File: build-system/eslint-rules/enums.js (hunk 1 above)
Comment (jridgewell) [reply build-system/eslint-rules/enums.js:91]: The comments are illustrative of what we're matching.

File: build-system/eslint-rules/enums.js [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            // name === 'enumValues' ||
+            // name === 'enumKeys' ||
+            // name === 'enumToObject'
+          ) {
```
Comment (jridgewell) [inline build-system/eslint-rules/enums.js:128]:
```suggestion
          if (name === 'isEnumValue') {
```

File: build-system/eslint-rules/enums.js [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+    context.report({
+      node,
+      message: [
+        `Improper use of enum, you may only do:`,
+        `- \`${node.name}.key\` get access.`,
+        `- \`isEnumValue(${node.name}, someValue)\` value checks.`,
+        // `- \`enumKeys(${node.name})\` keys gathering.`,
+        // `- \`enumValues(${node.name})\` values gathering.`,
+        // `- \`enumToObject(${node.name})\` conversion (AVOID if you can, it cannot be minified).`,
```
Comment (jridgewell) [inline build-system/eslint-rules/enums.js:143]:
```suggestion
```

File: build-system/eslint-rules/enums.js (hunk 2 above)
Comment (jridgewell) [reply build-system/eslint-rules/enums.js:124]: Done.

**Review by jridgewell — COMMENTED** [review]
`buid-system/` uses the funky Closure dialect of Typescript, so they do have `@enums` that are named `FOO_ENUM` and will not pass the linter. So I disabled.

**Review by samouri — APPROVED** [review]

**Pull request merged**